File: restring/src/main/java/com/ice/restring/MemoryStringRepository.java
package com.ice.restring;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * A StringRepository which keeps the strings ONLY in memory.
 * <p>
 * it's not ThreadSafe.
 */
class MemoryStringRepository implements StringRepository {

    private Map<String, Map<String, String>> strings = new LinkedHashMap<>();

    @Override
    public void setStrings(String language, Map<String, String> newStrings) {
        strings.put(language, newStrings);
    }

    @Override
    public void setString(String language, String key, String value) {
        if (!strings.containsKey(language)) {
            strings.put(language, new LinkedHashMap<>());
        }
        strings.get(language).put(key, value);
    }

    @Override
    public String getString(String language, String key) {
        if (!strings.containsKey(language) || !strings.get(language).containsKey(key)) {
            return null;
        }
        return strings.get(language).get(key);
    }

    @Override
    public Map<String, String> getStrings(String language) {
        if (!strings.containsKey(language)) {
            return new LinkedHashMap<>();
        }

        return new LinkedHashMap<>(strings.get(language));
    }
}

File: restring/src/main/java/com/b3nedikt/restring/repository/MemoryStringRepository.kt
package com.b3nedikt.restring.repository

import com.b3nedikt.restring.PluralKeyword
import com.b3nedikt.restring.StringRepository
import java.util.*

/**
 * A StringRepository which keeps the strings ONLY in memory.
 *
 *
 * it's not ThreadSafe.
 */
class MemoryStringRepository : StringRepository {

    private val strings = mutableMapOf<Locale, MutableMap<String, CharSequence>>()
    private val quantityStrings = mutableMapOf<Locale, MutableMap<String, Map<PluralKeyword, CharSequence>>>()
    private val stringArrays = mutableMapOf<Locale, MutableMap<String, Array<CharSequence>>>()

    override var supportedLocales: Set<Locale> = emptySet()

    override fun setStrings(locale: Locale, strings: Map<String, CharSequence>) {
        if (!this.strings.containsKey(locale)) {
            this.strings[locale] = mutableMapOf()
        }
        this.strings[locale] = strings.toMutableMap()
    }

    override fun setString(locale: Locale, key: String, value: CharSequence) {
        setStrings(locale, mapOf(key to value))
    }

    override fun getString(locale: Locale, key: String) = strings[locale]?.get(key)

    override fun getStrings(locale: Locale) = strings[locale]?.toMap() ?: mapOf()

    override fun getQuantityString(locale: Locale, key: String) = quantityStrings[locale]?.get(key)

    override fun setQuantityString(
            locale: Locale,
            key: String,
            quantityString: Map<PluralKeyword, CharSequence>) {
        setQuantityStrings(locale, mapOf(key to quantityString))
    }

    override fun getQuantityStrings(locale: Locale) = quantityStrings[locale]?.toMap() ?: mapOf()

    override fun setQuantityStrings(
            locale: Locale,
            quantityStrings: Map<String, Map<PluralKeyword, CharSequence>>) {
        if (!this.quantityStrings.containsKey(locale)) {
            this.quantityStrings[locale] = mutableMapOf()
        }
        this.quantityStrings[locale] = quantityStrings.toMutableMap()
    }

    override fun getStringArray(locale: Locale, key: String) = getStringArrays(locale)[key]

    override fun setStringArray(locale: Locale, key: String, stringArray: Array<CharSequence>) {
        setStringArrays(locale, mapOf(key to stringArray))
    }

    override fun getStringArrays(locale: Locale): Map<String, Array<CharSequence>> =
            stringArrays[locale] ?: mapOf()

    override fun setStringArrays(locale: Locale, stringArrays: Map<String, Array<CharSequence>>) {
        if (!this.stringArrays.containsKey(locale)) {
            this.stringArrays[locale] = mutableMapOf()
        }
        this.stringArrays[locale] = stringArrays.toMutableMap()
    }
}